lib/Sema/CSSimplify.cpp

```diff
@@ -694,13 +756,71 @@ static bool matchCallArgumentsImpl(
     // If we still haven't claimed all of the arguments,
     // fail if there is no recovery.
     if (numClaimedArgs != numArgs) {
+      bool unclaimedArgAfterCompletion = false;
       for (auto index : indices(claimedArgs)) {
         if (claimedArgs[index])
           continue;
 
+        // Avoid reporting individual extra arguments after the completion
+        // position. They are reported via a single callback below.
+        if (isAfterCompletionToken(completionInfo, index)) {
+          unclaimedArgAfterCompletion = true;
+          continue;
+        }
+
         if (listener.extraArgument(index))
           return true;
       }
+
+      if (completionInfo && unclaimedArgAfterCompletion) {
+        // There were invalid arguments after the completion position. The
+        // user may be intending to rewrite the callsite (rather than update an
+        // argument) so proceed as if all arguments after the completion
+        // position (even those already bound) weren't present. This prevents
+        // this solution being penalized based on the exact number and kind
+        // of issues those arguments contain.
+        unsigned afterIdx = completionInfo->completionIdx + 1;
+
+        auto argsBegin = args.begin() + afterIdx;
+        if (argsBegin != args.end())
+          args.erase(argsBegin, args.end());
+        numArgs = args.size();
+
+        if (!actualArgNames.empty()) {
+          auto actualNamesBegin = actualArgNames.begin() + afterIdx;
+          if (actualNamesBegin != actualArgNames.end())
+            actualArgNames.erase(actualNamesBegin, actualArgNames.end());
+        }
+
+        auto claimedBegin = claimedArgs.begin() + afterIdx;
+        if (claimedBegin != claimedArgs.end())
+          claimedArgs.erase(claimedBegin, claimedArgs.end());
+        numClaimedArgs = llvm::count_if(claimedArgs, [](bool claimed){
+          return claimed;
+        });
+
+        for (auto &params: llvm::reverse(parameterBindings)) {
+          if (params.empty())
+            continue;
+          if (!isAfterCompletionToken(completionInfo, params.back()))
+            break;
+          if (isAfterCompletionToken(completionInfo, params.front())) {
+            params.clear();
+            haveUnfulfilledParams = true;
+          } else {
+            llvm::erase_if(params, [&](unsigned argIdx) {
+              return isAfterCompletionToken(completionInfo, argIdx);
+            });
```

Comment:
Sorry, `params` was a terrible variable name. It should be `boundArgs` or something like that. It's the list of the argument indices bound to a specific param. The if condition above guarantees that in this branch the first argument index in the list isn't after the completion arg so there will always be at least one argument still bound to this param, so it can't become unfulfilled.  

That said, your comments about `break`ing above made me realise doing this whole thing isn't necessary anymore. The earlier changes in this function prevent invalid/out-of-order args after the completion position being bound and I've since updated the PR to handle any type mismatches in bound args after the completion elsewhere, so the only issue not stripping away the post-completion args causes now is some undesirable out-of-order/relabelling fixes in the `if (potentiallyOutOfOrder)` block below.  Those are only happening because we end up synthesising missing arguments for unfulfilled params and adding them to the end of `args`, and it ends up producing an out-of-order or relabelling fix if there are also non-synthesized extraneous args in the list (as they're before them). That's easy to fix with a check to see if the arg's index is >= numArgs though, so I'll just add that and delete this whole block.

Reply:
You could also update listener to do nothing for synthesis in code completion mode.